File: 10.py
test=0#1
import re
LINE = open(0).read().strip()

def p2(line: str):
    sequence = '17, 31, 73, 47, 23'.replace(' ','')
    line = ','.join([str(ord(c)) for c in line]) + ',' + sequence
    n = go(64,line,256)
    dh = []
    for i in range(0,len(n),16):
        x = n[i]
        for j in range(i+1,i+16):
            x ^= n[j]
        dh.append(hex(x)[2:].zfill(2))
    knot = ''.join(dh)
    assert len(knot) == 32
    return knot

def p1(line: str) -> int:
    n = go(1,line,256)
    return n[0]*n[1]

def go(n,line,RANGE):
    lens = getlens( line )
    nums = list(range(RANGE))
    if test==1:
        nums = list(range(5))
        lens = getlens('3, 4, 1, 5')
    skip = 0
    curr = 0
    assert not any([_ > RANGE for _ in lens])
    for _ in range(n):
        for ln in lens:
            indices,rev = [],[]
            for i in range(ln):
                idx = (curr + i) % RANGE
                indices.append(idx)
                rev.append(nums[idx])
            rev = rev[::-1]
            """
            print('curr/pos',curr,'/val',nums[curr],'\nskip/',skip)
            print('nums/',nums)
            print('idxs/',indices)
            print('rev/',rev)
            """
            for idx,i, in enumerate(indices):
                nums[i] = rev[idx]
            curr = (curr + ln + skip) % RANGE
            skip = (skip + 1) % RANGE
            """"
            print('inp/',ln, '\nnew/',nums)
            print('curr/pos',curr,'/val',nums[curr],'\nskip/',skip)
            print()
            """
    return nums

def getlens(s): return list(map(int,(re.findall(r'\d+', s))))

if __name__ == '__main__':
    print('part 1:', p1( LINE ))
    print('part 2:', p2( LINE ))

File: 10.cc
#include "iostream" // hex
#include "vector"
#include "sstream"
#include "algorithm" //reverse
#include "iomanip" // setfill setw

template <typename ... Params>
void print(const Params &... args) {
    ( (std::cout << args << ' '), ... ) << '\n';
}

std::vector<int> go(int, std::string, int);
std::string go2(std::string);

int main(){
    std::string line;
    std::getline(std::cin, line);
    print(line,"/line");
    std::vector<int> nums = go(1,line,256);
    std::string p2 = go2(line);

    print("part 1:", nums[0] * nums[1]);
    print("part 2:", go2(line));
}

std::string go2(std::string LINE) {
    std::string SEQ = "17, 31, 73, 47, 23";
    std::string line;
    for (char c: LINE)
        line += std::to_string(int(c)) + ',';
    for (char c: SEQ)
        if (c != ' ') line += c;
    std::vector<int> nums = go(64,line,256);
    int i = 0;
    int N = nums.size();
    std::string knot;
    while (i < N){
        int x = nums[i];
        int j = i;
        while (++j < i + 16)
            x ^= nums[j];
        // 2 ways to get a len-2 hex string
        //  1/ hex and iomanip
        std::stringstream ss;
        ss << std::hex << std::setw(2) << std::setfill('0') << x;
        knot += ss.str();
        //  2/ sprintf
        char hx[2];
        std::sprintf(hx, "%x", x);
        std::string res = hx;
        assert (ss.str() == (res.size() < 2 ? '0' + res : res));
        i += 16;
    }
    assert (knot.size() == 1<<5);
    return knot;
}

std::vector<int> go(int rounds, std::string line, int RANGE) {
    std::vector<int> lens;
    std::vector<int> nums(RANGE);
    std::stringstream ss(line);
    std::string tmp;
    while (std::getline(ss, tmp, ','))
        lens.push_back(stoi(tmp));
    int i = -1;
    while (++i < RANGE)
        nums[i] = i;
    int curr = 0;
    int skip = 0;
    i = -1;
    while (++i < rounds) {
        for (int ln: lens) {
            std::vector<int> indices, rv;
            int j = -1;
            while (++j < ln) {
                int idx = (curr + j) % RANGE;
                indices.push_back(idx);
                rv.push_back(nums[idx]);
            }
            std::reverse(rv.begin(), rv.end());
            j = -1;
            while (++j < indices.size())
                nums[indices[j]] = rv[j];
            curr = (curr + skip + ln) % RANGE;
            skip = (skip + 1) % RANGE;
        }
    }
    return nums;
}
